src/building/bootstrapping.md

```diff
@@ -122,50 +124,134 @@ contribution [here][bootstrap-build].
-Building LLVM for x86_64-apple-darwin
-Building stage0 codegen artifacts
-Assembling stage1 compiler
-Building stage1 std artifacts
-Copying stage1 std from stage1
-Building stage1 compiler artifacts
-Copying stage1 rustc from stage1
-Building stage1 codegen artifacts
-Assembling stage2 compiler
-Uplifting stage1 std
-Copying stage2 std from stage1
-Generating unstable book md files
-Building stage0 tool unstable-book-gen
-Building stage0 tool rustbook
-Documenting standalone
-Building rustdoc for stage2
-Documenting book redirect pages
-Documenting stage2 std
-Building rustdoc for stage1
-Documenting stage2 whitelisted compiler
-Documenting stage2 compiler
-Documenting stage2 rustdoc
-Documenting error index
-Uplifting stage1 rustc
-Copying stage2 rustc from stage1
-Building stage2 tool error_index_generator
-```
+### Overview
+
+This is a detailed look into the separate bootstrap stages.
+
+The convention `x.py` uses is that:
+- A `--stage N` flag means to run the stage N compiler (`stageN/rustc`).
+- A "stage N artifact" is a build artifact that is _produced_ by the stage N compiler.
+- The "stage (N+1) compiler" is assembled from "stage N artifacts". This
+  process is called _uplifting_.
+
+#### Build artifacts
+
+Anything you can build with `x.py` is a _build artifact_.
+Build artifacts include, but are not limited to:
+
+- binaries, like `stage0-rustc/rustc-main`
+- shared objects, like `stage0-sysroot/rustlib/libstd-6fae108520cf72fe.so`
+- [rlib] files, like `stage0-sysroot/rustlib/libstd-6fae108520cf72fe.rlib`
+- HTML files generated by rustdoc, like `doc/std`
+
+[rlib]: ../serialization.md
+
+#### Examples
+
+- `x.py build --stage 0` means to build with the beta `rustc`.
+- `x.py doc --stage 0` means to document using the beta `rustdoc`.
+- `x.py test --stage 0 library/std` means to run tests on the standard library
+    without building `rustc` from source ('build with stage 0, then test the
+  artifacts'). If you're working on the standard library, this is normally the
+  test command you want.
+- `x.py test src/test/ui` means to build the stage 1 compiler and run
+  `compiletest` on it. If you're working on the compiler, this is normally the
+  test command you want.
+
+#### Examples of what *not* to do
+
+- `x.py test --stage 0 src/test/ui` is not meaningful: it runs tests on the
+  _beta_ compiler and doesn't build `rustc` from source. Use `test src/test/ui`
+  instead, which builds stage 1 from source.
+- `x.py test --stage 0 compiler/rustc` builds the compiler but runs no tests:
+  it's running `cargo test -p rustc`, but cargo doesn't understand Rust's
+  tests. You shouldn't need to use this, use `test` instead (without arguments).
+- `x.py build --stage 0 compiler/rustc` builds the compiler, but does not make
+  it usable: the build artifacts are not uplifted ([#73519]). Use `x.py build
+  library/std` instead, which puts the compiler in `stage1/rustc`.
+
+[#73519]: https://github.com/rust-lang/rust/issues/73519
+
```

Comment:
What about having a table like this? (Correct me if I got things mixed up; it's confusing :) )

I think this will help clear things up for people (me).

| Stage | Flag for _building_ this stage | Flag for _running_ this stage | Output directory | What it's used for |
| --- | --- | --- | --- | --- |
| stage 0 | N/A | `--stage 0` | N/A | This is the latest beta version of rustc, similar to `rustc +beta ...` |
| stage 1 | `--stage 1` | `--stage 1` | `build/<toolchain>/stage1` | Usually you want to build this when working on Rust |
| stage 2 | `--stage 2` | `--stage 2` | `build<toolchain>/stage2` | This is what's released to rustup; you probably don't want to build this when contributing |

```suggestion

### Stages

| Stage | Flag for _building_ this stage | Flag for _running_ this stage | Output directory | What it's used for |
| --- | --- | --- | --- | --- |
| stage 0 | N/A | `--stage 0` | N/A | This is the latest beta version of rustc, similar to `rustc +beta ...` |
| stage 1 | `--stage 1` | `--stage 1` | `build/<toolchain>/stage1` | Usually you want to build this when working on Rust |
| stage 2 | `--stage 2` | `--stage 2` | `build<toolchain>/stage2` | This is what's released to rustup; you probably don't want to build this when contributing |

```

Reply:
Heh, this is basically https://github.com/rust-lang/rustc-dev-guide/pull/843. Let's hold off on that for now.

Reply:
Alright, maybe I'll open a follow-up PR then :)

Reply:
Also it's not entirely correct, `test --stage 1 src/test/ui` runs the stage1 compiler but `test --stage library/std` tests the stage1 _artifacts_.

Reply:
I'm not sure what you mean... I guess that's the point :)

Reply:
This difference between the compiler and the standard library is basically all the confusion I've seen about staging: sometimes --stage 1 means the compiler you're _running_ and other times it means the thing you _built_ and which is which depends on a whole bunch of different things. `test compiler/rustc` is the worst offender IMO but there's a lot making it confusing.

Reply:
The output directories are wrong too, it should by `stage1-<component>` where `<component>` is one of

- rustc
- std
- sysroot
- tools

`build/stage1` is the compiler itself, not the artifacts.

Reply:
FWIW there's already a table like this here (search 'Stage 0 action'): https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html?search=#understanding-stages-of-bootstrap

Reply:
It has a ton of detail though; I want something that's simple and gives the overview.

Reply:
I don't think we'll be able to get a simple chart; `--stage` means too many different things.

Reply:
Well, maybe not "simple" but more of an overview, not the details of what exactly is happening.

Reply:
I can add a link to 
![image](https://user-images.githubusercontent.com/23638587/94874258-924dc680-041f-11eb-8c90-b23948b513b3.png)
but you said yourself it confused you more than it helps. I don't know any better way to say it; any view that doesn't have different sections for the compiler and standard library is going to be incomplete and misleading.

Reply:
This table with the arrows is really helpful.
Could you add it to the `### Overview` section?

Reply:
There were a lot of people in https://github.com/rust-lang/rustc-dev-guide/pull/843 that said it confused more than it helps. I don't have the energy to push this through to an official document, I'm happy to host it on my blog or something.

Reply:
Or at least, if this is added I would like it to not be in this PR. This PR is mostly uncontroversial changes and gets rid of the wall of text and I don't want to lose those in a debate over what the model of bootstrapping should be.

Reply:
Oh, I haven't read #843. Nevermind then.